Clarify docs on different tximport count files (#1366)
nf-core/rnaseq: docs/output.md

  - `<pseudo_aligner>.merged.gene_counts.tsv`: Matrix of gene-level raw counts across all samples.
  - `<pseudo_aligner>.gene_tpm.tsv`: Matrix of gene-level TPM values across all samples.
  - `<pseudo_aligner>.gene_counts.rds`: RDS object that can be loaded in R that contains a [SummarizedExperiment](https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html) container with the TPM (`abundance`), estimated counts (`counts`) and transcript length (`length`) in the assays slot for genes.

>   - `<pseudo_aligner>.merged.gene_lengths.tsv`: Matrix of average within-sample transcript lengths for each gene across all samples, which can be used as an offset for gene-level analysis.

Reviewer: Same here, let's just be descriptive
```suggestion
  - `<pseudo_aligner>.merged.gene_lengths.tsv`: Matrix of average within-sample transcript lengths for each gene across all samples.
```

>   - `<pseudo_aligner>.merged.gene_counts_scaled.tsv`: Matrix of gene-level library size-scaled estimated counts from abundances across all samples.

Reviewer: Sorry, let's not bake the 'from abundances' in everywhere. I realise that's what the pseudo-aligners do, but it not necessary to spell that out here- the user just needs to know the content of the files.

Author: No worries; it's OK to be picky here. 

Just to get this straight though (in case I'm missing something): what I wanted to describe was that those files contain "counts", that were created by `tximport` adding back in length/library size information into the TPM values reported by e.g. `salmon`. 

I.e., from the tximport vignette:

> We could alternatively generate counts from abundances, using the argument countsFromAbundance, scaled to library size, "scaledTPM", or additionally scaled using the average transcript length, averaged over samples and to library size, "lengthScaledTPM". 

This as opposed to the gene_counts.tsv file, which contains gene-level aggregated counts (based on the transcript counts).

In any case, if you think it's clear as is, we can just omit those lines again.

Reviewer: As I understand it, abundance is effectively what these tools produce by default (via probabilistic maths I don't completely understand), such that even the content of `gene_counts.tsv` will be an estimate based on the abundance.

So I think it's best to keep things simple.

>   - `<pseudo_aligner>.merged.gene_counts_scaled.tsv`: Matrix of gene-level library size-scaled estimated counts from abundances across all samples.

Reviewer: ```suggestion
  - `<pseudo_aligner>.merged.gene_counts_scaled.tsv`: Matrix of gene-level library size-scaled estimated counts across all samples.
```

The [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html) package is used in this pipeline to summarise the results generated by Salmon or Kallisto into matrices for use with downstream differential analysis packages. We use tximport with different options to summarize count and TPM quantifications at the gene- and transcript-level. Please see [#499](https://github.com/nf-core/rnaseq/issues/499) for discussion and links regarding which counts are suitable for different types of analysis.

According to the [`txtimport` documentation](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#Downstream_DGE_in_Bioconductor) you can do one of the following:

> - Use the original (bias-uncorrected) counts _with an offset_: import all the salmon `quant.sf` files with `tximport` and then use `DESeq2` with `dds <- DESeqDataSetFromTximport(txi, sampleTable, ~condition)` to automatically correct for changes to the average transcript length across samples by calculating a gene-level offset.

Reviewer: ```suggestion
- Use the original (bias-uncorrected) counts _with an offset_: import all the salmon `quant.sf` files with `tximport` and then use `DESeq2` with `dds <- DESeqDataSetFromTximport(txi, sampleTable, ~condition)` to automatically construct an object with gene-level offsets in-built, which can be used by DESeq2 to automatically account for effective gene length effects across conditions. It's also possible to use the `.gene_lengths.tsv` matrices output by this workflow to make your own object the same way- which is what [the nf-core differentialabundance workflow does](https://nf-co.re/differentialabundance/1.5.0/docs/usage#outputs-from-nf-corernaseq-and-other-tximport-processed-results). See the [DESeq2 module](https://github.com/nf-core/modules/blob/c8f7f481bf2ccd8f9c7f2d499d94e74d4b9e23ab/modules/nf-core/deseq2/differential/templates/deseq_de.R#L323) for the implementation details.
```

Author: That looks even better, thanks! The description on the differential abundance page is also super helpful, so it's great that you linked to it.

I did notice that the `*.gene_lengths.tsv` file is missing from the list of output files above, so I'll go ahead and add it there.

Author: Oh, and when making the object yourself, you'd need to combine the counts with the lengths, wouldn't you? I.e.:

> It's also possible to use the `.merged.gene_counts.tsv` and  `.gene_lengths.tsv` matrices output by this workflow together....

Reviewer: Yep, the DESeq module link is there to illustrate how

> - Use bias-corrected counts _without an offset_: load and use `salmon.merged.gene_counts_length_scaled.tsv` or `salmon.merged.gene_counts_scaled.tsv` directly as you would with a regular gene-level counts matrix. These files were created using the `tximport` argument `countsFromAbundance="scaledTPM"` or `"lengthScaledTPM"` to scale counts to library size or to library size and average transcript length respectively, forgoing the need for an offset matrix.

Reviewer: ```suggestion
- Use bias-corrected counts _without an offset_: load and use `salmon.merged.gene_counts_length_scaled.tsv` or `salmon.merged.gene_counts_scaled.tsv` directly as you would with a regular gene-level counts matrix. These files were created using the `tximport` argument `countsFromAbundance="scaledTPM"` or `"lengthScaledTPM"` to scale counts to library size or to library size and average transcript length respectively.
```

Where possible, I think it's best to model the offsets explicitly, so I don't think we should imply that this a shortcut to an equivalent result.

Author: Sure, that seems to be the consensus I've read elsewhere too!
